### PR TITLE
I230 testing cloud functions

### PR DESCRIPTION
**Pull request by ondrejtomas**

### CODE REVIEW

File: ci/unit_tests/functions_test/main_test.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_invalidJsonInput(self):
+        ''' Tests if the input file contains invalid json '''
+        testArgs = [self.noJsonPath, self.noJsonPath] + self.functionsTestArgs
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'CRITICAL Cannot decode json from test input file ' + 
+            '/home/tomas/Projects/watson-assistant-workbench/ci/unit_tests/functions_test/main_data/no.json, ' +
```
Comment (ota-sandr) [inline ci/unit_tests/functions_test/main_test.py]: Personal path in test, nice :-P

File: ci/unit_tests/functions_test/main_test.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        )
+
+    def test_testSingleAllInFile(self):
+        ''' Tests if the single test where input and expected output is specified in file (all other params are given from command line) '''
+        outputFilePath = os.path.abspath(os.path.join(self.testOutputPath, os.path.splitext(os.path.basename(self.testSingleAllInFileJsonPath))[0] + '.out.json'))
+        testArgs = [self.testSingleAllInFileJsonPath, outputFilePath] + self.functionsTestArgs
+        self.t_noException([testArgs])
+        with open(outputFilePath, 'r') as outputFile:
+            outputJson = json.load(outputFile)
+            assert outputJson[0]['result'] == 0
```
Comment (ota-sandr) [inline ci/unit_tests/functions_test/main_test.py]: 1. You should check first if there is an array in `outputJson` and it has exactly one item
2. Consider adding `assert` which checks that there is no attribute `diff`

File: ci/unit_tests/functions_test/main_test.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    def test_nonArrayJsonInput(self):
+        ''' Tests if the input file contains non array json '''
+        testArgs = [self.emptyDictJsonPath, self.noJsonPath] + self.functionsTestArgs
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'CRITICAL Input test json is not array!', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_testSingleInvalid(self):
```
Comment (ota-sandr) [inline ci/unit_tests/functions_test/main_test.py]: I think that name `test_singleInvalid` would be sufficient (same for tests bellow).

File: ci/unit_tests/functions_test/main_test.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    def test_testSingleAllInFileOverrideFunction(self):
+        ''' Tests if bad package was provided in test it self, the main reason is test that override works '''
+        testArgs = [self.testSingleAllInFileOverrideFunctionJsonPath, self.noJsonPath] + self.functionsTestArgs
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'actions/utils/BAD_FUNCTION', # log message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_testSingleMulti(self):
```
Comment (ota-sandr) [inline ci/unit_tests/functions_test/main_test.py]: I think that correct name should be `test_multi`.

File: scripts/functions_test.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            ...
+            "outputReturned": <OBJECT>, # returned payload from CF
+            "result": <0 - test passed, 1 - test failed>
+            "diff": <OBJECT> # if test passed then "diff" is Null, else contains object that represents differences
+        }
+    ]
+    '''
+    logger.info('STARTING: '+ os.path.basename(__file__))
+    parser = argparse.ArgumentParser(description='Tests all tests specified in given file against Cloud Functions and save test outputs to output file', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    # positional arguments
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I would like to change / improve this for all script at once when we decide how our params should look like (I copied this from workspace_test.py).

File: scripts/functions_test.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+    # run tests
+    testCounter = 0
+    for test in inputJson:
+        if not isinstance(test, dict):
+            logger.critical('Input test array element %d is not dictionary. Each test has to be dictionary, please see doc!', testCounter)
+            sys.exit(1)
+        logger.info('Test number: %d, name: %s', testCounter, (test['name'] if 'name' in test else '-'))
+        testUrl = url + ('' if url.endswith('/') else '/') + namespace + '/actions/' + (test['cf_package'] if 'cf_package' in test else package) + '/' + (test['cf_function'] if 'cf_function' in test else function)
+        testUrl += '?blocking=true&result=true'
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        logger.info('Tested function url: %s', testUrl)
+
+        # load test input payload json
+        testInputJson = test['input']
+        if isinstance(testInputJson, basestring) and testInputJson.startswith('@'):
+            testInputRelativePath = testInputJson[1:]
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            logger.critical('Input test array element %d is not dictionary. Each test has to be dictionary, please see doc!', testCounter)
+            sys.exit(1)
+        logger.info('Test number: %d, name: %s', testCounter, (test['name'] if 'name' in test else '-'))
+        testUrl = url + ('' if url.endswith('/') else '/') + namespace + '/actions/' + (test['cf_package'] if 'cf_package' in test else package) + '/' + (test['cf_function'] if 'cf_function' in test else function)
+        testUrl += '?blocking=true&result=true'
+        logger.info('Tested function url: %s', testUrl)
+
+        # load test input payload json
+        testInputJson = test['input']
+        if isinstance(testInputJson, basestring) and testInputJson.startswith('@'):
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+                logger.critical('Cannot decode json from expected output payload from file %s, error: %s', testOutputExpectedPath, str(e))
+                sys.exit(1)
+
+        # call CF
+        logger.debug('Sending input json: %s', json.dumps(testInputJson, ensure_ascii=False).encode('utf8'))
+        response = requests.post(
+            testUrl, 
+            auth=(username, password), 
+            headers={'Content-Type': 'application/json'}, 
+            data=json.dumps(testInputJson, indent=4, ensure_ascii=False).encode('utf8'))
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+        responseContentType = response.headers.get('content-type')
+        if responseContentType != 'application/json':
+            logger.critical('Error while testing, response content type is not json, content type: %s, response:\n%s', responseContentType, response.text)
+            sys.exit(1)
+
+        # check status
+        if response.status_code in [200, 202]:
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                    test['result'] = 0
+                else:
+                    test['result'] = 1
+                    test['diff'] = testResultJson
+            else:
+                logger.critical('Error while evaluation, unknown test type: %s', test['type'])
+                sys.exit(1)
+        else:
+            logger.critical('Error while testing, response status: %d, response: %s', response.status_code, json.dumps(response.json(), ensure_ascii=False).encode('utf8'))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+        # check status
+        if response.status_code in [200, 202]:
+            testOutputReturnedJson = response.json()
+            logger.debug('Received output json: %s', json.dumps(testOutputReturnedJson, ensure_ascii=False).encode('utf8'))
+            test['outputReturned'] = testOutputReturnedJson
+
+            # evaluate test
+            if 'type' not in test or test['type'] == 'EXACT_MATCH':
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: Future extensions.

File: scripts/functions_test.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+'''
+
+import json, sys, os, argparse, requests, configparser
+from wawCommons import setLoggerConfig, getScriptLogger, getRequiredParameter, getOptionalParameter
+from cfgCommons import Cfg
+import logging
+from deepdiff import DeepDiff
+from pprint import pformat
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply scripts/functions_test.py]: I'll improve it.

File: ci/unit_tests/functions_test/main_test.py [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+"""
+
+import os, pytest, json
+
+import functions_test
+from ...test_utils import BaseTestCaseCapture
+
+class TestMain(BaseTestCaseCapture):
+
+    dataBasePath = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'main_data' + os.sep)
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: I'll improve it.

File: ci/unit_tests/functions_test/main_test.py [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+    def test_args_basic(self):
+        ''' Tests some basic sets of args '''
+        self.t_tooFewArgs([[]])
+        self.t_tooFewArgs([['/some/random/path']])
+        self.t_unrecognizedArgs([['/some/random/path', '-s', 'randomNonPositionalArg']])
+
+    def test_args_advanced(self):
+        ''' Tests cloud function sets of args '''
+        for missingTestArgIndex, missingTestArg in enumerate(self.functionsTestArgs[::2]):
+            print(missingTestArg)
```
[the remark this answers is not in the document]
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: I'll improve it.

File: ci/unit_tests/functions_test/main_test.py (hunk 1 above)
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: I'll improve it.

File: ci/unit_tests/functions_test/main_test.py (hunk 2 above)
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: I'll improve it.

File: ci/unit_tests/functions_test/main_test.py (hunk 3 above)
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: The second 'test' in the name means that in the input test file there is only one test and not more.

File: ci/unit_tests/functions_test/main_test.py (hunk 4 above)
Comment (ondrejtomas) [reply ci/unit_tests/functions_test/main_test.py]: I'll improve it.

File: scripts/functions_test.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            url + ('' if url.endswith('/') else '/') + \
+            namespace + '/actions/' + (test['cf_package'] if 'cf_package' in test else package) + '/' + \
+            (test['cf_function'] if 'cf_function' in test else function) + \
+            '?blocking=true&result=true'
+        logger.info('Tested function url: %s', testUrl)
+
+        # load test input payload json
+        testInputJson = test['input']
+        try:
+            if testInputJson.startswith('@'):
```
Comment (ota-sandr) [inline scripts/functions_test.py]: You did not have `else` clause for this `if` (same for code bellow)

File: ci/test_utils.py [hunk 17]
```diff
@@ -81,58 +81,106 @@ def t_fun_exitCodeAndErrMessage(self, function, exitCode, errMessage, args=[], k
...
-        elif messageType == BaseTestCaseCapture.MessageType.ERR:
-            assert logMessage in self.captured.err
-        elif messageType == BaseTestCaseCapture.MessageType.LOG:
-            assert logMessage in self.logs.text
-        else:
-            pytest.fail('Uknown MessageType: ' + messageType)
+        self.t_fun_generic(function, SystemExit, str(exitCode), messageType, message, args, kwargs)
 
-    def t_raiseError(self, errorType, errMessage, args=[], kwargs={}):
+    def t_raiseError(self, exceptionType, exceptionValue, args=[], kwargs={}):
```
Comment (ota-sandr) [inline ci/test_utils.py]: Consider renaming to `t_raiseException` (because you already renamed parameters).

**Pull request merged**